### ISSUE TITLE
Source-to-source AD: wrong-sign gradient on cross-component advection kernels (~1.6% fd_check error)

### ISSUE THREAD

**Issue by andrinr** [NVIDIA/warp#1472]
### Bug Description

**Update after a clean repro attempt: I cannot reproduce this on Warp 1.13.0, on either CPU or CUDA, for either the 2-D or the 3-D variant of the kernel.** Leaving the issue open in case maintainers recognise the failure mode from an older release, but it's likely either fixed in 1.13.0 or only reachable under conditions my isolated repro doesn't capture (e.g. multi-step graphs, specific input distributions, JIT interactions). I'll close if no one recognises it.

The original claim was: Warp's source-to-source AD on a tentative-velocity step for incompressible Navier–Stokes (`u* = u + dt·(-u·∇u + ν∇²u)`) produced wrong-sign gradients on the cross-component advection terms, with `fd_check rel_err ~1.6%` and cosine similarity against an analytic reference adjoint plateauing around 0.973. The downstream workaround was to replace the auto-adjoint with an explicit numpy backward registered via `wp.Tape.record_func`.

### Repro attempt and result

Self-contained script comparing three adjoints of the 2-D kernel: Warp's tape, central finite difference, and an analytic adjoint hand-derived from the kernel.

```python
import numpy as np
import warp as wp

device = "cuda:0"   # also tried "cpu" — same result

@wp.kernel
def tentative_vel_2d(
    ux: wp.array2d(dtype=wp.float32), uy: wp.array2d(dtype=wp.float32),
    ux_star: wp.array2d(dtype=wp.float32), uy_star: wp.array2d(dtype=wp.float32),
    dt: float, inv_2h: float, inv_h2: float, nu: float,
):
    i, j = wp.tid()
    n = ux.shape[0]
    ip1 = (i + 1) % n; im1 = (i - 1 + n) % n
    jp1 = (j + 1) % n; jm1 = (j - 1 + n) % n
    ui = ux[i, j]; vi = uy[i, j]
    lap_ux = (ux[im1, j] + ux[ip1, j] + ux[i, jm1] + ux[i, jp1] - 4.0 * ui) * inv_h2
    adv_ux = ui * (ux[ip1, j] - ux[im1, j]) * inv_2h + vi * (ux[i, jp1] - ux[i, jm1]) * inv_2h
    ux_star[i, j] = ui + dt * (-adv_ux + nu * lap_ux)
    lap_uy = (uy[im1, j] + uy[ip1, j] + uy[i, jm1] + uy[i, jp1] - 4.0 * vi) * inv_h2
    adv_uy = ui * (uy[ip1, j] - uy[im1, j]) * inv_2h + vi * (uy[i, jp1] - uy[i, jm1]) * inv_2h
    uy_star[i, j] = vi + dt * (-adv_uy + nu * lap_uy)

# ... (build random ux, uy; run tape.backward; compare to FD adjoint)
```

Results on Warp 1.13.0, V100 (sm_70), float32, N = 64:

```
adj_ux:
  warp     vs FD       : rel_err=2.8e-5   cos_sim=1.000000
  analytic vs FD       : rel_err=2.8e-5   cos_sim=1.000000
  warp     vs analytic : rel_err=8.6e-8   cos_sim=1.000000

adj_uy:
  warp     vs FD       : rel_err=2.8e-5   cos_sim=1.000000
  analytic vs FD       : rel_err=2.8e-5   cos_sim=1.000000
  warp     vs analytic : rel_err=8.2e-8   cos_sim=1.000000
```

The 2.8e-5 floor is the FP32 central-difference accuracy; Warp's tape matches the analytic adjoint to ~1e-7. Same pattern for the 3-D variant at N = 8 (~2.8e-5 vs FD across all three components).

### Possible explanations

- The bug existed in an earlier Warp release and was fixed before 1.13.0.
- The failure was triggered by a specific upstream condition I'm not reproducing in isolation: a particular input distribution, a long pipeline of subsequent kernels accumulating wrong adjoints, or an interaction with `requires_grad` ordering elsewhere in the program.
- The downstream observation was misdiagnosed and the real bug was somewhere else in the graph.

If you've seen anything like the failure mode described in the original report, I'm happy to dig further; otherwise I'll close this.

### System Information

- Warp 1.13.0, CUDA 12.9, NVIDIA Tesla V100-PCIE-16GB (sm_70), Python 3.10, x86_64 Linux

**Comment by andrinr** [comment]
Confirmed not reproducible in 1.13.0 (CPU and CUDA, 2D and 3D). The symptom that motivated the report came from a different source in our pipeline, not from Warp's auto-adjoint. Closing.
